File: numerix/numerix/src/pkg/middleware/middleware.rs
use std::{future::Future, mem::replace, pin::Pin, task::{Context, Poll}, time::Instant};
use hyper::{ Request, Response};
use tower::{Layer, Service};
use std::time::Duration;
use hyper::StatusCode;
use crate::pkg::metrics::metrics;
use crate::logger;
use std::fmt;

#[derive(Debug, Default, Clone)]
pub struct GrpcMiddlewareLayer;

impl<S> Layer<S> for GrpcMiddlewareLayer {
    type Service = GrpcMiddlewareService<S>;

    fn layer(&self, service: S) -> Self::Service {
        GrpcMiddlewareService { inner : service }
    }
}

#[derive(Debug, Clone)]
pub struct GrpcMiddlewareService<S> {
    inner : S,
}

type BoxFuture<'a, T> = Pin<Box<dyn Future<Output = T> + Send + 'a>>;

impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for GrpcMiddlewareService<S> 
where
    S: Service<Request<ReqBody>, Response = Response<ResBody>> + Clone + Send + 'static,
    S::Future: Send + 'static,
    ReqBody: Send + 'static,
    S::Error: fmt::Display + fmt::Debug + Send + Sync + 'static,
{
    type Response = S::Response;
    type Error = S::Error;
    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.inner.poll_ready(cx)
    }

    fn call(&mut self, req: Request<ReqBody>) -> Self::Future
    {
        let clone = self.inner.clone();
        let mut inner = replace(&mut self.inner, clone);

        Box::pin(async move {
            let start = Instant::now();
            let method = req.method().clone();
                        
            let response = inner.call(req).await;
            
            let status_code = match &response {
                Ok(resp) => resp.status(),
                Err(_) => StatusCode::INTERNAL_SERVER_ERROR,
            };
        
            let duration = start.elapsed();

            

            let response = match response {
                Ok(resp) => {
                    resp
                },
                Err(e) => {
                    logger::error(
                        format!("Request failed: method={}, status={}, duration={:?}, error={}", 
                            method, status_code, duration, e), 
                        None
                    );
                    return Err(e);
                }
            };

            telemetry_middleware(method.as_str(), duration, status_code);

            Ok(response)
        })
    }
}

fn telemetry_middleware(method: &str, duration: Duration, status_code: StatusCode)  {
    let tags = vec![("api", method), ("status", status_code.as_str())];
    
    let _ = metrics::timing("router.api.request.latency", duration, &tags);
    let _ = metrics::count("router.api.request.total", 1, &tags);
}

File: inferflow/pkg/middleware/middleware.go
//go:build !meesho

package middleware

import (
	"context"
	"encoding/json"
	"runtime/debug"
	"strconv"
	"strings"
	"time"

	"github.com/rs/zerolog/log"

	"github.com/Meesho/BharatMLStack/inferflow/pkg/logger"
	"github.com/Meesho/BharatMLStack/inferflow/pkg/metrics"
	"github.com/Meesho/BharatMLStack/inferflow/pkg/set"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
)

var (
	reqHeadersToLog *set.ThreadSafeSet
)

func InitGRPCMiddleware() {
	reqHeadersToLog = set.NewThreadSafeSet()
}

func WrappedGRPCMiddleware(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo,
	handler grpc.UnaryHandler) (resp interface{}, err error) {
	startTime := time.Now()

	// Extract metadata from the context
	md, ok := metadata.FromIncomingContext(ctx)
	if !ok {
		return nil, status.Errorf(codes.InvalidArgument, "Retrieving metadata is failed")
	}

	if err := authorize(md); err != nil {
		return nil, err
	}
	// Extract relevant information from the request
	method := info.FullMethod
	requestHeaders, _ := json.Marshal(filterGRPCHeaders(md))

	// Call the gRPC handler to handle the request
	resp, err = handler(ctx, req)
	statusCode := codes.OK
	if err != nil {
		statusCode = status.Code(err)
	}
	// Extract relevant information from the response
	responseTime := time.Since(startTime)

	// Log the request and response information
	logVariables := []string{
		method,
		strconv.Itoa(int(statusCode)),
		responseTime.String(),
		string(requestHeaders),
	}
	if err != nil {
		logger.Error(strings.Join(logVariables, " | "), err)
	} else {
		logger.Info(strings.Join(logVariables, " | "))
	}
	telemetryMiddleware(info, responseTime, statusCode)
	return resp, err
}

func RecoveryInterceptor(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (resp interface{}, err error) {
	defer func() {
		if r := recover(); r != nil {
			// Log the panic, stack trace, and the request
			log.Error().
				Str("service", info.FullMethod).
				Interface("request", req).
				Msgf("Recovered in recovery interceptor with err: %v, stack: %s", r, string(debug.Stack()))
			// Convert the panic to a gRPC error
			err = status.Errorf(codes.Internal, "Internal server error")
		}
	}()

	// Call the handler
	return handler(ctx, req)
}

func authorize(md metadata.MD) error {
	// Add authorization logic here if needed
	return nil
}

func filterGRPCHeaders(md metadata.MD) map[string][]string {

	filteredHeaders := make(map[string][]string)
	for k, v := range md {
		if reqHeadersToLog.Contains(k) {
			filteredHeaders[k] = v
		}
	}
	return filteredHeaders
}

func telemetryMiddleware(info *grpc.UnaryServerInfo, responseTime time.Duration, statusCode codes.Code) {
	tags := []string{"api:" + info.FullMethod, "status:" + strconv.Itoa(int(statusCode))}
	metrics.Timing("inferflow.router.api.request.latency", responseTime, tags)
	metrics.Count("inferflow.router.api.request.total", 1, tags)
}
